src/libsyntax/parse/parser.rs

```diff
@@ -8712,22 +8713,43 @@ impl<'a> Parser<'a> {
     ///
     /// The arguments of the function are replaced in HIR lowering with the arguments created by
     /// this function and the statements created here are inserted at the top of the closure body.
-    fn construct_async_arguments(&mut self, asyncness: &mut Spanned<IsAsync>, decl: &FnDecl) {
+    fn construct_async_arguments(&mut self, asyncness: &mut Spanned<IsAsync>, decl: &mut FnDecl) {
+        struct PatVisitor {
+            mutability: Mutability,
+        }
+        impl<'a> Visitor<'a> for PatVisitor {
```

Comment:
It seems that this could be avoided if we made `move_local` optional instead of `pat_stmt`. Then we can have `move_local` always create a mutable binding when it exists.

Reply:
Thanks for pointing that out! It was used because I had overlooked that the following pattern didn't receive an `unused_mut` warning. So I removed it.  

```rust
fn foo(__arg0: A) {
    let mut __arg0 = __arg0;
    let ref x = __arg0;
}
```

Reply:
Wait, it doesn't? How come? Also, I hope we're generating something more like this:
```rust
fn foo(mut __arg0: A) {
    let ref x = __arg0;
}
```
(Unless that has the wrong drop order or something)

Reply:
We don't lint for variables with names starting with `_`

Reply:
Oops, I thought the `ref` was causing the change in behavior, my bad!